### ISSUE TITLE
Receiving JSON data 

### ISSUE THREAD

**Issue by connerpaulge** [miguelgrinberg/Flask-SocketIO#556]
Hi all

So the premise of what I am doing is that I am sending JSON (correctly formatted) data from a server to port 80 on my AWS EC2 web server and then socket.io will take the data and continuously display the data to a highcharts graph.  It is running Nginx + gunicorn + gevent with websockets.

I am able to get the socket.io connection to get a 200 but no data ever displays and I cant figure out why.

Here is my code:

On the server side it is listening for json:

@socketio.on('json')
def handle_json(json):
    print('received json: ' + str(json))

On the client side it should be logging the json to the console:
socket.on('json', function(json) {
                        console.log(json)
                });

Any help is appreciated!  If I am totally doing this the wrong way let me know of a different path to try and I'll give it a shot.

Thanks
Conner

**Comment by connerpaulge** [comment]
Adding on to this issue:

I have tried setting up a fake page with flask "/processjson"  where I have flask handle the json and return some value of that json.  

I set up flask like this:


@app.route('/processjson', methods=['GET','POST'])
def processjson():
   data = request.get_json()

    ID = data['RasID']
    return jsonify({'result' : 'Success', 'RasID' : ID })

and then I used a program called Postman to send a post request to that page and it returned success with the value I passed in.

Basically this is telling me that the server IS receiving the data and it can be used but flask-socketio doesn't seem to want to receive it.

Why is that?

Thanks
Conner

**Comment by miguelgrinberg** [comment]
@connerpaulge aren't you missing an `emit()` somewhere? Based on your description I think on your EC2 you would need to call `socketio.emit('json', json)` at the place you receive the json data. This will send the data over Socket.IO to your client-side application, which can then feed it to your charts.

**Comment by connerpaulge** [comment]
Miguel

I am a little confused by your comment.  Let me explain how I thought of it (I am likely wrong). 

I thought that if I am sending JSON data to, let's just say a fake IP of 152.25.25.5, and in the __init__.py in the flask config I have the server listening for JSON data (see below), then the __init__.py would receive the data and then print it right out to the console.  That was my first testing step, to actually try and get the data to print out.  

 **@socketio.on('json')
def handle_json(json):
print('received json: ' + str(json))**


My next step is trying to have the specific page, let's say 152.25.25.5/dashboard, actually log the data to the specific page's console with the below code.  So I had thought that if the server receives the data in the first step above and prints it to the console I should then send / emit the data to all the pages, which if the page has this code below it should then log the JSON data in the console.

**socket.on('json', function(json) {
console.log(json)
});**

Please correct me if I am wrong but that is how I had assumed it would work.

Thank you for all the help!!!
Conner

**Comment by connerpaulge** [comment]
Update:

I think I understand your comment a little bit better.

But I still am having trouble understanding why this isnt working.(see below)  I had thought that code snippet below would handle and emit the json to the client-side application??  It still doesn't even print anything out.  I am confused to why it doesn't even print out and JSON

**@socketio.on('json')
def handle_json(json):
    print('received json: ' + str(json))
    socketio.emit('json', json)
    send(json, broadcast=True)**

**Comment by miguelgrinberg** [comment]
@connerpaulge you haven't shown any piece of code that sends JSON data to your script. How does that work?

**Comment by connerpaulge** [comment]
I thought (send(json, broadcast=True)) would broadcast the json data to anything which was listening and the code (see below) would listen for json and then log it to the console?  Or am I totally wrong?

socket.on('json', function(json) {
console.log(json)
});

**Comment by miguelgrinberg** [comment]
> I thought (send(json, broadcast=True)) would broadcast the json data to anything which was listening

That is correct. But you have this call inside the `handle_json()` function. Who calls that function?

**Comment by connerpaulge** [comment]
OHHHHHHH I get it now.  I fixed it for it to call the handle_json() and it works.  

Miguel honestly thank you so much!!

**Comment by miguelgrinberg** [comment]
Sure, no problem. But note that if you are calling this function manually, then you don't need the `@socketio.on` decorator added to it. That is only for functions that are invoked as a result of an event.
